### ISSUE TITLE
salt-cp randomly not working

### ISSUE THREAD

**Issue by g3kk0** [saltstack/salt#35454]
## Description

When using salt-cp to push a file to a minion it randomly fails to work.

The issue appears to relate to the size of the file being pushed. Any file size greater than 250KB seems to cause the problem and results in about half of all copy attempts failing.
## Setup

`salt-cp '*' /tmp/file.txt /tmp`

Note: single master -> single minion
### Broken

```
# salt-cp '*' -l debug /tmp/file.txt /tmp
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Configuration file path: /etc/salt/master
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Missing configuration file: /root/.saltrc
[DEBUG   ] MasterEvent PUB socket URI: /var/run/salt/master/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: /var/run/salt/master/master_event_pull.ipc
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'salt.example.com_master', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] Initializing new IPCClient for path: /var/run/salt/master/master_event_pub.ipc
[DEBUG   ] LazyLoaded local_cache.get_load
[DEBUG   ] Reading minion list from /var/cache/salt/master/jobs/35/5cee972e8fdc0c5325f6d23f526921/.minions.p
[DEBUG   ] get_iter_returns for jid 20160815174112095876 sent to set(['dev-web-1']) will timeout at 17:41:17.374448
[DEBUG   ] Checking whether jid 20160815174112095876 is still running
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'salt.example.com_master', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] LazyLoaded nested.output
```

```
# /var/log/salt/master
2016-08-15 17:41:11,919 [salt.utils.verify][WARNING ][31110] Insecure logging configuration detected! Sensitive data may be logged.
2016-08-15 17:41:11,920 [salt.config      ][DEBUG   ][31110] Reading configuration from /etc/salt/master
2016-08-15 17:41:12,050 [salt.config      ][DEBUG   ][31110] Missing configuration file: /root/.saltrc
2016-08-15 17:41:12,052 [salt.utils.event ][DEBUG   ][31110] MasterEvent PUB socket URI: /var/run/salt/master/master_event_pub.ipc
2016-08-15 17:41:12,052 [salt.utils.event ][DEBUG   ][31110] MasterEvent PULL socket URI: /var/run/salt/master/master_event_pull.ipc
2016-08-15 17:41:12,088 [salt.transport.zeromq][DEBUG   ][31110] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'salt.example.com_master', 'tcp://127.0.0.1:4506', 'clear')
2016-08-15 17:41:12,089 [salt.transport.ipc][DEBUG   ][31110] Initializing new IPCClient for path: /var/run/salt/master/master_event_pub.ipc
2016-08-15 17:41:12,372 [salt.utils.lazy  ][DEBUG   ][31110] LazyLoaded local_cache.get_load
2016-08-15 17:41:12,374 [salt.loaded.int.returner.local_cache][DEBUG   ][31110] Reading minion list from /var/cache/salt/master/jobs/35/5cee972e8fdc0c5325f6d23f526921/.minions.p
2016-08-15 17:41:12,375 [salt.client      ][DEBUG   ][31110] get_iter_returns for jid 20160815174112095876 sent to set(['dev-web-1']) will timeout at 17:41:17.374448
2016-08-15 17:41:17,382 [salt.client      ][DEBUG   ][31110] Checking whether jid 20160815174112095876 is still running
2016-08-15 17:41:17,383 [salt.transport.zeromq][DEBUG   ][31110] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'salt.example.com_master', 'tcp://127.0.0.1:4506', 'clear')
2016-08-15 17:41:27,483 [salt.utils.lazy  ][DEBUG   ][31110] LazyLoaded nested.output
```
### Successful

```
# salt-cp '*' -l debug /tmp/file.txt /tmp
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Configuration file path: /etc/salt/master
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Missing configuration file: /root/.saltrc
[DEBUG   ] MasterEvent PUB socket URI: /var/run/salt/master/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: /var/run/salt/master/master_event_pull.ipc
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'salt.example.com_master', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] Initializing new IPCClient for path: /var/run/salt/master/master_event_pub.ipc
[DEBUG   ] LazyLoaded local_cache.get_load
[DEBUG   ] Reading minion list from /var/cache/salt/master/jobs/38/66afa23fc3b01fbecdc54f5c262f5c/.minions.p
[DEBUG   ] get_iter_returns for jid 20160815173927931889 sent to set(['dev-web-1']) will timeout at 17:39:33.333354
[DEBUG   ] jid 20160815173927931889 return from dev-web-1
[DEBUG   ] jid 20160815173927931889 found all minions set(['dev-web-1'])
[DEBUG   ] LazyLoaded nested.output
dev-web-1:
    ----------
    /tmp/file.txt:
        True
```

```
# /var/log/salt/master
2016-08-15 17:40:54,852 [salt.utils.verify][WARNING ][31051] Insecure logging configuration detected! Sensitive data may be logged.
2016-08-15 17:40:54,853 [salt.config      ][DEBUG   ][31051] Reading configuration from /etc/salt/master
2016-08-15 17:40:54,982 [salt.config      ][DEBUG   ][31051] Missing configuration file: /root/.saltrc
2016-08-15 17:40:54,983 [salt.utils.event ][DEBUG   ][31051] MasterEvent PUB socket URI: /var/run/salt/master/master_event_pub.ipc
2016-08-15 17:40:54,984 [salt.utils.event ][DEBUG   ][31051] MasterEvent PULL socket URI: /var/run/salt/master/master_event_pull.ipc
2016-08-15 17:40:55,019 [salt.transport.zeromq][DEBUG   ][31051] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'salt.example.com_master', 'tcp://127.0.0.1:4506', 'clear')
2016-08-15 17:40:55,020 [salt.transport.ipc][DEBUG   ][31051] Initializing new IPCClient for path: /var/run/salt/master/master_event_pub.ipc
2016-08-15 17:40:55,302 [salt.utils.lazy  ][DEBUG   ][31051] LazyLoaded local_cache.get_load
2016-08-15 17:40:55,303 [salt.loaded.int.returner.local_cache][DEBUG   ][31051] Reading minion list from /var/cache/salt/master/jobs/ae/aa5fb581a5725eb44b8d16054a410e/.minions.p
2016-08-15 17:40:55,304 [salt.client      ][DEBUG   ][31051] get_iter_returns for jid 20160815174055028577 sent to set(['dev-web-1']) will timeout at 17:41:00.303705
2016-08-15 17:40:55,909 [salt.client      ][DEBUG   ][31051] jid 20160815174055028577 return from dev-web-1
2016-08-15 17:40:56,010 [salt.client      ][DEBUG   ][31051] jid 20160815174055028577 found all minions set(['dev-web-1'])
2016-08-15 17:40:56,012 [salt.utils.lazy  ][DEBUG   ][31051] LazyLoaded nested.output
```
## Versions

```
Salt Version:
           Salt: 2016.3.2

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.5 (default, Nov 20 2015, 02:00:19)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.2.1511 Core
        machine: x86_64
        release: 3.10.0-327.22.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.2.1511 Core
```

**Comment by Ch3LL** [comment]
@g3kk0 I am able to replicate this. I am also seeing this on 2015.8 branch as well. 

1. `fallocate -l 500KB big_file`
2. `fallocat -l 1K small_file`
3. small file copy:

```
[ch3ll@thecakeisalie ~]$ sudo salt-cp 'ch3ll-minion02' /home/ch3ll/small_file /tmp/
ch3ll-minion02:
    ----------
    /tmp/small_file:
        True
```

4. big file copy

```
[ch3ll@thecakeisalie ~]$ sudo salt-cp 'ch3ll-minion02' /home/ch3ll/big_file /tmp/ -ldebug
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: thecakeisalie
[DEBUG   ] Configuration file path: /etc/salt/master
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: thecakeisalie
[DEBUG   ] Missing configuration file: /root/.saltrc
[DEBUG   ] MasterEvent PUB socket URI: ipc:///var/run/salt/master/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: ipc:///var/run/salt/master/master_event_pull.ipc
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'thecakeisalie_master', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] LazyLoaded local_cache.get_load
[DEBUG   ] get_iter_returns for jid 20160815152421709373 sent to set(['ch3ll-minion02']) will timeout at 15:24:27.403020
[DEBUG   ] Checking whether jid 20160815152421709373 is still running
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'thecakeisalie_master', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] LazyLoaded nested.output
```

and it stops there. Looks like we need to fix this up.

**Comment by thatch45** [comment]
@g3kk0 salt-cp we never intended to copy large files and frankly, I am surprised that it was able to handle more than 50K!
salt-cp works by sending the file down via the salt pub message, which is typically around 256 bytes, but with salt-cp it gets greatly ballooned.
So this is a long standing issue, if you want to transfer large files you will need to use the cp.get_file function: (https://docs.saltstack.com/en/2015.8/ref/modules/all/salt.modules.cp.html#salt.modules.cp.get_file)

There is a long standing idea to change salt-cp to use the fileserver backend(salt-cp was built before salt had a file server, or even config management), but we have never prioritized it due to a lack of demand. The idea for it is to make salt-cp run cp.getfile under the hood and to move a copy of the file to be transferred into the salt fileserver. This of course also has issues since the salt fileserver can use a gitfs backend now and disable the local fileserver.

So, long story short, if you need to transfer large files please use cp.get_file

**Comment by g3kk0** [comment]
Thanks for the prompt response. The switch to cp.get_file works as expected.

During my investigation I did come across a post which hinted salt-cp was only for smaller files but couldn't find anything official. I see you've already created a request to update the docs which is great.

As you say it would be great to hook this into salt-cp therefore avoiding the extra steps required in getting the files to the fileserver directory.

**Comment by thatch45** [comment]
Yes, I could have sworn I had note in the docs about it, but I imagine that they were lost over the years, it happens :)
